### ISSUE TITLE
fzf inappropriate ioctl for device

### ISSUE THREAD

**Issue by roginfarrer** [mikew/nvrh#12]
Hello! This is a really cool project. I've been stuck using VSCode for work, so I was excited to see you post it on Reddit.

I just started exploring it, but I quickly ran into an error when trying to run fzf-lua. I got the following error:

```
[Fzf-lua] 'fzf' is not a valid executable, aborting. Please make sure 'fzf' is in installed.
fzf inappropriate ioctl for device
```

While this is an fzf error, I wonder if this is related to how nvrh connects to remote? I have fzf installed on both the client and the server, so I'm not sure exactly why this doesn't work.

**Comment by mikew** [comment]
Should be something to do with the server-side portion, since with nvrh there's nothing running on the client outside of neovim itself.

What's your `$PATH` like, when in neovim via nvrh? Opening a terminal and doing `echo "$PATH"` might hint at something.

**Comment by mikew** [comment]
For reference, my remote is Linux where I use linuxbrew (so for me, the `fzf` binary won't be found unless my shell RC files are loaded), but it seems to work?


https://github.com/user-attachments/assets/b00ae7e4-8c38-46a2-877c-6110cfc6769b

**Comment by roginfarrer** [comment]
I think you're onto something with the PATH. I get a different result from `!echo $PATH` and `lua print(vim.env.PATH)` (the latter lacking the path to my fzf binary on remote). Interesting that neovim initializes with a different PATH.

I use zsh on the server (a CentOS 7 machine). From some limitations at work, I have to use zsh, and usually run a startup check that if it's in interactive mode, go into fish shell. But I get the same result if I change it to stay in zsh.

**Comment by mikew** [comment]
Hm, that video I posted above was using zsh on the remote side of things.

What happens if you run `ssh vscode-remote -t '$SHELL -i -c "which fzf"'`. Hoping if that prints the path to fzf, it's something to do with nvim config.

**Comment by roginfarrer** [comment]
It didn't have fzf in the path as a result of that command. What I did try was manually add the folders for fzf my `.profile` and `.zprofile`, and so when I ssh into my server normally, they are in my PATH. (I specify them in my fish config files, so zsh wasn't adding them manually.)

So currently I ssh into my server, run `echo $PATH`, and I see the added directories. If I run `ssh YOUR_SERVER -t '$SHELL -i -c "which fzf"'`, I do not see the added directories. If I run `ssh YOUR_SERVER -t '$SHELL -i -c "source $ZDOTDIR/.zprofile && which fzf"'` source $ZDOTDIR/.zprofile && echo $PATH`, it does show executable.

When I place an `echo ".zprofile"` in my `.zprofile`, it does get echoed without sourcing it directly.

**Comment by mikew** [comment]
Hm, I can add a flag to let people specify extra commands to run before nvrh starts nvim. Would that work for you? I personally don't use .zprofile / most .z* files, just a .zshrc

**Comment by roginfarrer** [comment]
After some reading, it looks like `.zprofile` is only sourced at log-in, and not for new terminal sessions. I'm not 100% sure why that doesn't work in this case, but moving my `PATH` updates to `.zshrc` does fix the issue! I'm cool with making that change, so I think that should fix my issues.

I am curious about why you call `$SHELL -i -c` here, instead of the nvim command immediately?

https://github.com/mikew/nvrh/blob/59061b9b997e1235c497d855fd5bbeceaf8b9f6c/src/ssh_helpers/main.go#L24

**Comment by mikew** [comment]
> I am curious about why you call $SHELL -i -c here

Yeah absolutely! Kinda the the thing we're talking about lol. It's so people's shell configs are loaded.

As mentioned, I use linuxbrew. Its `bin` folder is added to my `PATH` in my `.zshrc`. If I run this, `fzf` isn't found:

```sh
ssh vscode-remote 'which fzf'
fzf not found
```

Starting `nvim` in a similar fashion wouldn't be great. It wouldn't know where any of my tools are.

However, using `$SHELL -i` gets your users default shell started in interactive mode (which loads all of your shell's RC files):

```sh
$ ssh -t vscode-remote '$SHELL -i -c "which fzf"'
/home/linuxbrew/.linuxbrew/bin/fzf
```

The `-c` is just telling it to run a command.

---

TLDR, it's a way to launch your shell in interactive mode, which tells your shell to load all its config files.
